### PR TITLE
Add sub-action to merge artifacts

### PR DESCRIPTION
**Pull request by robherley**
Follow up to:
- https://github.com/actions/upload-artifact/pull/504

Part of:
- https://github.com/actions/upload-artifact/pull/503

This introduces a new sub-action, `actions/upload-artifact/merge@v4` that facilitates merging one or more artifacts from a workflow.

For instance, the following is now possible:

```yaml
jobs:
  upload:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        foo: [a, b, c]

    steps:
      - name: Run a one-line script
        run: echo "hello from job ${{ matrix.foo }}" > file-${{ matrix.foo }}.txt
      - name: Upload
        uses: actions/upload-artifact@v4
        with:
          name: my-artifact-${{ matrix.foo }}
          path: file-${{ matrix.foo }}.txt
  merge:
    runs-on: ubuntu-latest
    needs: upload

    steps:
      - name: Merge Artifacts
        uses: actions/upload-artifact/merge@v4
        with:
          name: my-merged-artifact
          pattern: my-artifact-*
```

Where all the contents of `my-artifact-a`, `my-artifact-b`, `my-artifact-c` will be in a single artifact called `my-merged-artifact`.

### CODE REVIEW

File: .github/workflows/test.yml
Comment (bethanyj28) [inline .github/workflows/test.yml:1]: Could we get a test for this case?
```yaml
jobs:
  # ... <upload job> ...
  merge:
    runs-on: ubuntu-latest
    needs: upload
    steps:
      - name: Merge Artifacts
        uses: actions/upload-artifact/merge@v4
```

**Review by konradpabjan — APPROVED** [review]

**Review by bethanyj28 — COMMENTED** [review]
Looks good, just one question!

File: .github/workflows/test.yml
Comment (robherley) [reply .github/workflows/test.yml:1]: added 😎

**Review by bethanyj28 — APPROVED** [review]
🚀

**Pull request merged**